db/migrate/20190618202817_add_user_id_to_tag.rb

```diff
@@ -0,0 +1,79 @@
+class AddUserIdToTag < ActiveRecord::Migration[5.2]
+  def self.up
+    add_column :tags, :user_id, :integer
+
+    # Find uses of each tag for both Todos and RecurringTodos to
+    # figure out which users use which tags.
+    @tags = execute <<-EOQ
+      SELECT t.id AS tid, tds.user_id AS todo_uid, rt.user_id AS rtodo_uid
+      FROM tags t
+      JOIN taggings tgs ON tgs.tag_id = t.id
+      LEFT OUTER JOIN todos tds
+       ON tgs.taggable_type = "Todo" AND tds.id = tgs.taggable_id
+      LEFT OUTER JOIN recurring_todos rt
+       ON tgs.taggable_type = "RecurringTodo" AND rt.id = tgs.taggable_id
+      WHERE rt.id IS NOT NULL OR tds.id IS NOT NULL
+      GROUP BY t.id, tds.user_id, rt.user_id
+    EOQ
+
+    # Map each tag to the users using it.
+    @tag_users = {}
+    @tags.each do |row|
+      uid = (row[1] ? row[1] : row[2])
+      if not @tag_users[row[0]]
+        @tag_users[row[0]] = [uid]
+      elsif not @tag_users[row[0]].include? uid
+        @tag_users[row[0]] << uid
+      end
+    end
+
+    # Go through the tags assigning users and duplicating as necessary.
+    @tag_users.each do |tid, uids|
+      tag = Tag.find(tid)
+
+      # One of the users will get the original tag instance, but first
+      # duplicate their own copy to all the others.
+      extras = uids.length - 1
+      extras.times do |n|
+        uid = uids[n+1]
+
+        # Create a duplicate of the tag assigned to the user.
+        new_tag = tag.dup
+        new_tag.user_id = uid
+        new_tag.save!
+
+        # Move all the user's regular todos to the new tag.
+        execute <<-EOQ
```

Comment:
It's not a very good practice to inject variables inside the SQL statement without quoting, but since a migration is a limited scope and all the values are obtained from the same database, I think it's okay. The main problem was that I failed to find a working way for doing a proper prepared statement with placeholders here.

Reply:
I just found [`exec_update`](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/DatabaseStatements.html#method-i-exec_update) and we can use that for prepared statement type things, rather than using a raw `execute` here.

Reply:
I tried it, but I failed to figure out a working syntax with more than one variable. I spent some hours trying to figure it out, if someone else knows how it works feel free to change it.